### PR TITLE
Issue #2269 Add --no-proxy flag to oc-env

### PR DESCRIPTION
**Pull request by svenstaro**
The test is currently failing because I can't figure out how to get a `libmachine.API` to pass into it. Help appreciated.

### CODE REVIEW

File: cmd/minishift/cmd/oc_env.go [hunk 1]
```diff
@@ -58,10 +61,36 @@ func getOcShellConfig(ocPath, forcedShell string) (*OcShellConfig, error) {
 	cmdLine := "minishift oc-env"
 	shellCfg := &OcShellConfig{
 		OcDirPath: filepath.Dir(ocPath),
-		UsageHint: shell.GenerateUsageHint(userShell, cmdLine),
 	}
 
-	shellCfg.Prefix, shellCfg.Suffix, shellCfg.Delimiter = shell.GetPrefixSuffixDelimiterForSet(userShell, true)
+	if noProxy {
```
Comment (svenstaro) [inline cmd/minishift/cmd/oc_env.go]: Will get rid of this duplicated code with a function.

File: cmd/minishift/cmd/oc_env_test.go [hunk 2]
```diff
@@ -25,7 +25,7 @@ import (
 )
 
 func Test_unix_oc_path(t *testing.T) {
-	shellConfig, err := getOcShellConfig("/Users/john/.minishift/cache/oc/v1.5.0/oc", "bash")
+	shellConfig, err := getOcShellConfig("/Users/john/.minishift/cache/oc/v1.5.0/oc", "bash", false)
```
Comment (svenstaro) [inline cmd/minishift/cmd/oc_env_test.go]: This is what fails now because I can't get a `libmachine.API` object in a test.

File: cmd/minishift/cmd/oc_env.go (hunk 1 above)
Comment (praveenkumar) [reply cmd/minishift/cmd/oc_env.go]: Having this as part of method will make sense so this can be used for docker env and oc env both.

Comment (svenstaro) [reply cmd/minishift/cmd/oc_env.go]: Done.

File: cmd/minishift/cmd/util/util.go [hunk 3]
```diff
@@ -112,3 +114,28 @@ func IsValidProfileName(profileName string) bool {
 	rr := regexp.MustCompile(`^[a-zA-Z0-9]+[a-zA-Z0-9-]*$`)
 	return rr.MatchString(profileName)
 }
+
+func GetNoProxyConfig(api libmachine.API) (string, string, error) {
```
Comment (praveenkumar) [inline cmd/minishift/cmd/util/util.go]: :+1:  Good that now it separated and can be consumed by any command.

File: cmd/minishift/cmd/oc_env.go [hunk 4]
```diff
@@ -84,13 +96,15 @@ var ocEnvCmd = &cobra.Command{
...
 		util.ExitIfUndefined(api, constants.MachineName)
-		_, err := api.Load(constants.MachineName)
+
+		host, err := api.Load(constants.MachineName)
 		if err != nil {
 			atexit.ExitWithMessage(1, err.Error())
 		}
 
-		var shellCfg *OcShellConfig
+		util.ExitIfNotRunning(host.Driver, constants.MachineName)
```
Comment (praveenkumar) [inline cmd/minishift/cmd/oc_env.go]: :+1:

File: cmd/minishift/cmd/oc_env.go [hunk 5]
```diff
@@ -99,7 +113,7 @@ var ocEnvCmd = &cobra.Command{
 		} else {
 			ocPath = config.InstanceConfig.OcPath
 		}
-		shellCfg, err = getOcShellConfig(ocPath, forceShell)
+		shellCfg, err = getOcShellConfig(api, ocPath, forceShell, noProxy)
```
Comment (praveenkumar) [inline cmd/minishift/cmd/oc_env.go]: :+1:

File: cmd/minishift/cmd/oc_env_test.go [hunk 6]
```diff
@@ -21,11 +21,15 @@ package cmd
 import (
 	"testing"
 
+	"github.com/docker/machine/libmachine"
 	"github.com/stretchr/testify/assert"
 )
 
 func Test_unix_oc_path(t *testing.T) {
-	shellConfig, err := getOcShellConfig("/Users/john/.minishift/cache/oc/v1.5.0/oc", "bash")
+	api := libmachine.NewClient("foo", "foo")
```
Comment (praveenkumar) [inline cmd/minishift/cmd/oc_env_test.go]: As of now this is a dummy call so that we can pass it to under test function.

**Comment by praveenkumar** [comment]
Tested, works as expected.

```
$ ./minishift oc-env --no-proxy
export PATH="/home/prkumar/.minishift/cache/oc/v3.9.0/linux:$PATH"
export NO_PROXY="192.168.42.83"
# Run this command to configure your shell:
# eval $(minishift oc-env --no-proxy)
```

File: cmd/minishift/cmd/util/util.go (hunk 3 above)
Comment (LalatenduMohanty) [reply cmd/minishift/cmd/util/util.go]: 👍

**Comment by gbraad** [comment]
retest this please

**Comment by LalatenduMohanty** [comment]
retest this please

**Comment by LalatenduMohanty** [comment]
Integration tests failed with below error
```
 --- Failed steps:
18:07:39 
18:07:39   Scenario: Setting oc binary to PATH with oc-env command # features/cmd-oc-env.feature:19
18:07:39     Then executing "oc status" in host shell succeeds # features/cmd-oc-env.feature:22
18:07:39       Error: Command 'oc status', expected to succeed, exited with exit code: 127
18:07:39 
```
I manually checked and did not see the issue.

```
```
Scenario: Setting oc binary to PATH with oc-env command
     When executing "minishift oc-env" in host shell succeeds
      And evaluating stdout of the previous command in host shell
     Then executing "oc status" in host shell succeeds
      And stdout of host shell contains "In project My Project (myproject)"
```
$ minishift oc-env
export PATH="/Users/lmohanty/.minishift/cache/oc/v3.9.0/darwin:$PATH"
# Run this command to configure your shell:
# eval $(minishift oc-env)

$ eval $(minishift oc-env)

$ oc status
In project My Project (myproject) on server https://192.168.64.10:8443

You have no services, deployment configs, or build configs.
Run 'oc new-app' to create an application.

**Comment by praveenkumar** [comment]
@agajdosi With this PR looks like something wrong with `ExecuteInHostShellLineByLine()` I did local testing with the PR after adding log.

```patch
diff --git a/Makefile b/Makefile
index bc436f4f..8b028f3a 100644
--- a/Makefile
+++ b/Makefile
@@ -200,7 +200,7 @@ test: vendor $(ADDON_ASSET_FILE)  ## Run unit tests
        @go test -v -tags "$(BUILD_TAGS)" -ldflags="$(VERSION_VARIABLES)" $(shell $(PACKAGES))
 
 .PHONY: integration
-integration: GODOG_OPTS = --tags=basic ## Run integration tests
+integration: GODOG_OPTS = --tags=cmd-oc-env ## Run integration tests
 integration: $(MINISHIFT_BINARY)
        mkdir -p $(INTEGRATION_TEST_DIR)
        go test -timeout $(TIMEOUT) $(REPOPATH)/test/integration --tags=integration -v -args --test-dir $(INTEGRATION_TEST_DIR) --binary $(MINISHIFT_BINARY) \
diff --git a/test/integration/util/shell.go b/test/integration/util/shell.go
index 1c3a1900..bec36f69 100644
--- a/test/integration/util/shell.go
+++ b/test/integration/util/shell.go
@@ -29,6 +29,7 @@ import (
        "time"
 
        "github.com/DATA-DOG/godog/gherkin"
+       "github.com/docker/machine/libmachine/log"
        "github.com/minishift/minishift/pkg/util/os"
 )
 
@@ -271,6 +272,7 @@ func ExecuteInHostShellLineByLine() error {
        commandArray := strings.Split(stdout, "\n")
        for index := range commandArray {
                if !strings.Contains(commandArray[index], exitCodeIdentifier) {
+                       log.Errorf("\n\n Command line: %s\n\n", commandArray[index])
                        err = ExecuteInHostShell(commandArray[index])
                }
        }
```

This looks following on this PR which is wrong.

```
Scenario: Starting minishift                 # features/cmd-oc-env.feature:13
    Given Minishift has state "Does Not Exist" # testsuite.go:76 -> *Minishift./testsuite.shouldHaveState
    And image caching is disabled              # testsuite.go:91 -> *Minishift./testsuite.setImageCaching
    When executing "minishift start" succeeds  # testsuite.go:593 -> /testsuite.executingMinishiftCommandSucceedsOrFails
    Then Minishift has state "Running"         # testsuite.go:76 -> *Minishift./testsuite.shouldHaveState

  Scenario: Setting oc binary to PATH with oc-env command                 # features/cmd-oc-env.feature:19
    When executing "minishift oc-env" in host shell succeeds              # shell.go:288 -> /util.ExecuteMinishiftInHostShellSucceedsOrFails


 Command line: Running this command requires a running 'minishift' VM, but no VM is running. 




 Command line: 


    And evaluating stdout of the previous command in host shell           # shell.go:269 -> /util.ExecuteInHostShellLineByLine
    Then executing "oc status" in host shell succeeds                     # shell.go:237 -> /util.ExecuteInHostShellSucceedsOrFails
    Command 'oc status', expected to succeed, exited with exit code: 127

    And stdout of host shell contains "In project My Project (myproject)" # shell.go:297 -> /util.HostShellCommandReturnShouldContain

  Scenario: Deleting Minishift                         # features/cmd-oc-env.feature:25
    Given user closes shell instance on host machine   # shell.go:204 -> /util.CloseHostShellInstance
    And Minishift has state "Running"                  # testsuite.go:76 -> *Minishift./testsuite.shouldHaveState
    When executing "minishift delete --force" succeeds # testsuite.go:593 -> /testsuite.executingMinishiftCommandSucceedsOrFails
    Then Minishift has state "Does Not Exist"         
```

**Comment by agajdosi** [comment]
@LalatenduMohanty, @praveenkumar 
So the issue seems to be in L:10 of cmd-oc-env.feature, where the PATH is unset in the shell instance, thus docker-machine-driver-kvm is not accessible. We then start the Minishift through the Go's exec package, where driver is present and so it starts fine. After that the feature tries to run `minishift oc-env` and fails, because the driver is not on the path of the shell instance and thus it fails to get the M status.

As fix we might just remove `Scenario: User starts shell instance without oc in PATH`, the test will still make sense, the scenario only secured cases when tester has oc on path already. However we will follow-up issue in which I will implement the unset again but only for the oc binary using some regex parsing.

The feature until now worked because in the Master we have probably a bug which caused `oc-env` to not fail when docker-machine-driver-kvm was not present on path:
```
[agajdosi@dhcp-24-156 minishift]$ /home/agajdosi/golang/bin/minishift status
Minishift:  Running
Profile:    minishift
OpenShift:  Running (openshift v3.9.0+4814d4a-3)
DiskUsage:  9% of 17.9G
[agajdosi@dhcp-24-156 minishift]$ mv docker-machine-driver-kvm docker-machine-driver-kvm-222
[agajdosi@dhcp-24-156 minishift]$ /home/agajdosi/golang/bin/minishift oc-env
export PATH="/home/agajdosi/.minishift/cache/oc/v3.9.0/linux:$PATH"
# Run this command to configure your shell:
# eval $(minishift oc-env)
```
However this only works when M is running:
```
[agajdosi@dhcp-24-156 minishift]$ mv docker-machine-driver-kvm-222 docker-machine-driver-kvm
[agajdosi@dhcp-24-156 minishift]$ /home/agajdosi/golang/bin/minishift delete
You are deleting the Minishift VM: 'minishift'. Do you want to continue [y/N]?: y
Deleting the Minishift VM...
Minishift VM deleted.
[agajdosi@dhcp-24-156 minishift]$ mv docker-machine-driver-kvm docker-machine-driver-kvm-222
[agajdosi@dhcp-24-156 minishift]$ /home/agajdosi/golang/bin/minishift oc-env
Cannot find the OpenShift client binary.
Make sure that OpenShift was provisioned successfully.
```


Going to create an issue for this edge-case bug, which somehow got fixed in this PR, but it might probably need more deep checks in the code.

**Comment by praveenkumar** [comment]
@svenstaro Can you please remove step https://github.com/minishift/minishift/blob/master/test/integration/features/cmd-oc-env.feature#L10 and again push your changes to make sure CI passes.

> The feature until now worked because in the Master we have probably a bug which caused oc-env to not fail when docker-machine-driver-kvm was not present on path

@agajdosi It was not the bug but since we didn't had `util.ExitIfNotRunning(host.Driver, constants.MachineName)` as part of implementation before but now it is.

**Comment by praveenkumar** [comment]
retest this please.

File: pkg/util/shell/shell.go [hunk 7]
```diff
@@ -88,40 +89,32 @@ func GenerateUsageHint(userShell, cmdLine string) string {
...
-func GetPrefixSuffixDelimiterForSet(userShell string, pathVar bool) (prefix, suffix, delimiter string) {
+func GetPrefixSuffixDelimiterForSet(userShell string) (prefix, delimiter, suffix, pathSuffix string) {
 	switch userShell {
 	case "fish":
 		prefix = "set -gx "
-		suffix = "\";\n"
 		delimiter = " \""
-		if pathVar {
-			suffix = "\" $PATH;\n"
-		}
```
Comment (LalatenduMohanty) [inline pkg/util/shell/shell.go]: @svenstaro @praveenkumar Why is this change required? i.e. changes to `pathvar`

File: test/integration/features/cmd-oc-env.feature [hunk 8]
```diff
@@ -7,7 +7,6 @@ one from: bash, cmd, powershell, tcsh, zsh with TEST_WITH_SPECIFIED_SHELL parame
   Scenario: User starts shell instance without oc in PATH
   INFO: This scenario starts interactive shell instance, which will be closed in the end of this feature.
     Given user starts shell instance on host machine
-     When executing "unset PATH" in host shell
```
[the remark this answers is not in the document]
Comment (gbraad) [reply test/integration/features/cmd-oc-env.feature]: @svenstaro so the failure on CI was related to the test/feature ? ;-)

File: pkg/util/shell/shell.go (hunk 7 above)
Comment (budhrg) [reply pkg/util/shell/shell.go]: @LalatenduMohanty @svenstaro introduced a new member `PathSuffix` in struct `ShellConfig` and managing the current form of building path having `PATH` with it. Nice one! :+1:

**Comment by LalatenduMohanty** [comment]
> The feature until now worked because in the Master we have probably a bug which caused oc-env to not fail when docker-machine-driver-kvm was not present on path:

@agajdosi Are you sure about this because, 

```
$ /home/agajdosi/golang/bin/minishift oc-env
Cannot find the OpenShift client binary.
Make sure that OpenShift was provisioned successfully.
```
this above error is there from log time.

**Comment by agajdosi** [comment]
@LalatenduMohanty When M is running and machine-driver is not on Path, it works - which is not expected I think. When M is not running and machine-driver is or is not on Path, it does not work - which is expected.
